From: Duncan Eastoe <duncan@graphiant.com>
Date: Mon, 29 Jul 2024 15:13:17 +0100
Subject: ly_wrap: discard ignored libyang errors

If specific libyang errors are ignored, discard them. Otherwise they
are effectively leaked until/unless a unignored error is encountered.
---
 src/ly_wrap.c | 48 +++++++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 43 insertions(+), 5 deletions(-)

diff --git a/src/ly_wrap.c b/src/ly_wrap.c
index fa56a5f..039a2be 100644
--- a/src/ly_wrap.c
+++ b/src/ly_wrap.c
@@ -434,6 +434,8 @@ sr_lys_find_xpath_atoms(const struct ly_ctx *ctx, const char *xpath, uint32_t op
         } else if (lyrc == LY_ENOTFOUND) {
             /* no error message */
             sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, NULL);
+            /* free any errors which were generated */
+            ly_err_clean((struct ly_ctx *) ctx, NULL);
         } else {
             sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         }
@@ -488,6 +490,9 @@ sr_lyd_parse_data(const struct ly_ctx *ctx, const char *data, const char *data_p
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    ly_err_clean((struct ly_ctx *) ctx, NULL);
+
 cleanup:
     if (err_info) {
         lyd_free_siblings(*tree);
@@ -793,6 +798,9 @@ sr_lyd_new_meta2(const struct ly_ctx *ctx, struct lyd_node *parent, const struct
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    ly_err_clean((struct ly_ctx *) ctx, NULL);
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
@@ -1030,6 +1038,7 @@ sr_lyd_find_path(const struct lyd_node *tree, const char *path, int with_incompl
     sr_error_info_t *err_info = NULL;
     uint32_t temp_lo = LY_LOSTORE;
     LY_ERR lyrc;
+    const struct ly_ctx *ctx = tree ? LYD_CTX(tree) : NULL;
 
     ly_temp_log_options(&temp_lo);
 
@@ -1039,10 +1048,15 @@ sr_lyd_find_path(const struct lyd_node *tree, const char *path, int with_incompl
             *match = NULL;
         }
     } else if (lyrc && (lyrc != LY_ENOTFOUND)) {
-        sr_errinfo_new_ly(&err_info, LYD_CTX(tree), NULL, SR_ERR_LY);
+        sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    if (ctx) {
+        ly_err_clean((struct ly_ctx *) ctx, NULL);
+    }
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
@@ -1054,15 +1068,21 @@ sr_lyd_find_sibling_first(const struct lyd_node *sibling, const struct lyd_node
     sr_error_info_t *err_info = NULL;
     uint32_t temp_lo = LY_LOSTORE;
     LY_ERR lyrc;
+    const struct ly_ctx *ctx = sibling ? LYD_CTX(sibling) : NULL;
 
     ly_temp_log_options(&temp_lo);
 
     lyrc = lyd_find_sibling_first(sibling, target, match);
     if (lyrc && (lyrc != LY_ENOTFOUND)) {
-        sr_errinfo_new_ly(&err_info, LYD_CTX(sibling), NULL, SR_ERR_LY);
+        sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    if (ctx) {
+        ly_err_clean((struct ly_ctx *) ctx, NULL);
+    }
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
@@ -1075,15 +1095,21 @@ sr_lyd_find_sibling_val(const struct lyd_node *sibling, const struct lysc_node *
     sr_error_info_t *err_info = NULL;
     uint32_t temp_lo = LY_LOSTORE;
     LY_ERR lyrc;
+    const struct ly_ctx *ctx = sibling ? LYD_CTX(sibling) : NULL;
 
     ly_temp_log_options(&temp_lo);
 
     lyrc = lyd_find_sibling_val(sibling, schema, value, value ? strlen(value) : 0, match);
     if (lyrc && (lyrc != LY_ENOTFOUND)) {
-        sr_errinfo_new_ly(&err_info, LYD_CTX(sibling), NULL, SR_ERR_LY);
+        sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    if (ctx) {
+        ly_err_clean((struct ly_ctx *) ctx, NULL);
+    }
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
@@ -1095,15 +1121,21 @@ sr_lyd_find_sibling_opaq_next(const struct lyd_node *sibling, const char *name,
     sr_error_info_t *err_info = NULL;
     uint32_t temp_lo = LY_LOSTORE;
     LY_ERR lyrc;
+    const struct ly_ctx *ctx = sibling ? LYD_CTX(sibling) : NULL;
 
     ly_temp_log_options(&temp_lo);
 
     lyrc = lyd_find_sibling_opaq_next(sibling, name, match);
     if (lyrc && (lyrc != LY_ENOTFOUND)) {
-        sr_errinfo_new_ly(&err_info, LYD_CTX(sibling), NULL, SR_ERR_LY);
+        sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    if (ctx) {
+        ly_err_clean((struct ly_ctx *) ctx, NULL);
+    }
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
@@ -1151,15 +1183,21 @@ sr_lyd_change_term(struct lyd_node *node, const char *value, int ignore_fail)
     sr_error_info_t *err_info = NULL;
     uint32_t temp_lo = LY_LOSTORE;
     LY_ERR lyrc;
+    const struct ly_ctx *ctx = node ? LYD_CTX(node) : NULL;
 
     ly_temp_log_options(&temp_lo);
 
     lyrc = lyd_change_term(node, value);
     if (lyrc && (!ignore_fail || ((lyrc != LY_EEXIST) && (lyrc != LY_ENOT)))) {
-        sr_errinfo_new_ly(&err_info, LYD_CTX(node), NULL, SR_ERR_LY);
+        sr_errinfo_new_ly(&err_info, ctx, NULL, SR_ERR_LY);
         goto cleanup;
     }
 
+    /* discard any suppressed errors */
+    if (ctx) {
+        ly_err_clean((struct ly_ctx *) ctx, NULL);
+    }
+
 cleanup:
     ly_temp_log_options(NULL);
     return err_info;
